http2 INTERNAL_ERROR when combining vanguard with httputil.ReverseProxy

For reasons outside my control, I have a server that only serves grpc-web code and a client that only supprts grpc. I was looking to combine vanguard with reverse proxy to bridge the two. 

However, when trying to combine a vanguard transcoder with `httputil.NewSingleHostReverseProxy`, I'm running into some unexpected stream resets being presented to the client as `INTERNAL_ERROR`. Repro steps:

```go
package bug

import (
	"context"
	"crypto/tls"
	"fmt"
	"net"
	"net/http"
	"net/http/httptest"
	"net/http/httputil"
	"net/url"
	"testing"

	"buf.build/gen/go/connectrpc/eliza/connectrpc/go/connectrpc/eliza/v1/elizav1connect"
	elizav1 "buf.build/gen/go/connectrpc/eliza/protocolbuffers/go/connectrpc/eliza/v1"
	"connectrpc.com/connect"
	"connectrpc.com/vanguard"
	"github.com/stretchr/testify/require"
	"golang.org/x/net/http2"
	"golang.org/x/net/http2/h2c"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/reflect/protoregistry"
)

func TestINTERNAL_ERROR(t *testing.T) {

	// GRPC-web server (normally process 1)
	mux := http.NewServeMux()
	mux.Handle(elizav1connect.NewElizaServiceHandler(&handler{t: t}))
	grpcwebServer := httptest.NewUnstartedServer(
		h2c.NewHandler(mux, &http2.Server{}),
	)
	grpcwebServer.EnableHTTP2 = true
	grpcwebServer.Start()
	defer grpcwebServer.Close()

	// Vanguard based proxy (normally process 2)
	url, err := url.Parse(grpcwebServer.URL)
	require.NoError(t, err)
	proxy := httputil.NewSingleHostReverseProxy(url)
	proxy.FlushInterval = -1 // shouldn't be necessary
	// Normally I load these from a file, but since everything is in the same process for the test grab them here
	schema, err := protoregistry.GlobalFiles.FindDescriptorByName("connectrpc.eliza.v1.ElizaService")
	require.NoError(t, err)
	transcoder, err := vanguard.NewTranscoder(
		[]*vanguard.Service{vanguard.NewServiceWithSchema(schema.(protoreflect.ServiceDescriptor), proxy)},
		vanguard.WithDefaultServiceOptions(
			vanguard.WithTargetProtocols(vanguard.ProtocolGRPCWeb),
			vanguard.WithTargetCodecs(vanguard.CodecProto),
		),
		vanguard.WithUnknownHandler(proxy),
	)
	require.NoError(t, err)
	vanguardServer := httptest.NewUnstartedServer(
		h2c.NewHandler(
			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				// https://cs.opensource.google/go/go/+/master:src/net/http/httputil/reverseproxy.go;l=520;drc=960654be0c4ad7918376e2e1d47491c9bc7520e0
				// https://github.com/connectrpc/vanguard-go/blob/c6d89f4b3bbda0b6fa4737a9bb2c4425c4a23bfa/transcoder.go#L1530
				// https://github.com/connectrpc/vanguard-go/blob/c6d89f4b3bbda0b6fa4737a9bb2c4425c4a23bfa/transcoder.go#L1329-L1362
				//
				// defer func() {
				// 	if r := recover(); r != nil {
				// 		if r != http.ErrAbortHandler {
				// 			panic(r)
				// 		}
				// 	}
				// }()
				t.Log("vanguard server: ", r.Header)
				transcoder.ServeHTTP(w, r)
			}),
			&http2.Server{},
		),
	)
	vanguardServer.EnableHTTP2 = true
	vanguardServer.Start()
	defer vanguardServer.Close()

	// grpc client using h2c (normally process 3)
	h2cClient := &http.Client{
		Transport: &http2.Transport{
			AllowHTTP: true,
			DialTLS: func(network, addr string, cfg *tls.Config) (net.Conn, error) {
				return net.Dial(network, addr)
			},
		},
	}
	client := elizav1connect.NewElizaServiceClient(h2cClient, vanguardServer.URL, connect.WithGRPC(), connect.WithProtoJSON())
	unaryResp, err := client.Say(context.Background(), connect.NewRequest(&elizav1.SayRequest{Sentence: "foo"}))
	require.NoError(t, err)
	require.Equal(t, unaryResp.Msg.Sentence, "echo foo")

	serverStream, err := client.Introduce(context.Background(), connect.NewRequest(&elizav1.IntroduceRequest{Name: "foo"}))
	require.NoError(t, err)
	var responses []string
	for serverStream.Receive() {
		responses = append(responses, serverStream.Msg().Sentence)
	}
	require.NoError(t, serverStream.Err())
	require.Equal(t, []string{"hi", "hi again"}, responses)
}

type handler struct {
	elizav1connect.UnimplementedElizaServiceHandler
	t testing.TB
}

func (h *handler) Say(ctx context.Context, req *connect.Request[elizav1.SayRequest]) (*connect.Response[elizav1.SayResponse], error) {
	if req.Peer().Protocol != connect.ProtocolGRPCWeb {
		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("only accepts grpc-web, got %q", req.Peer().Protocol))
	}
	h.t.Log("connect handler: ", req.Header())
	return connect.NewResponse(&elizav1.SayResponse{Sentence: "echo " + req.Msg.Sentence}), nil
}

func (h *handler) Introduce(ctx context.Context, req *connect.Request[elizav1.IntroduceRequest], resp *connect.ServerStream[elizav1.IntroduceResponse]) error {
	if req.Peer().Protocol != connect.ProtocolGRPCWeb {
		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("only accepts grpc-web, got %q", req.Peer().Protocol))
	}
	resp.Send(&elizav1.IntroduceResponse{Sentence: "hi"})
	resp.Send(&elizav1.IntroduceResponse{Sentence: "hi again"})
	return nil
}
```


```
$ GODEBUG=http2debug=1 go test . -v
2024/11/03 14:48:21 http2: Transport received RST_STREAM stream=3 len=4 ErrCode=INTERNAL_ERROR
    vanguard_test.go:88: 
                Error Trace:    /Users/robbert/Desktop/vanguard_test.go:88
                Error:          Received unexpected error:
                                internal: stream error: stream ID 3; INTERNAL_ERROR; received from peer
                Test:           TestINTERNAL_ERROR
--- FAIL: TestINTERNAL_ERROR (0.00s)
FAIL
FAIL    bug     0.011s
FAIL
```

From my peek into this 
https://cs.opensource.google/go/go/+/master:src/net/http/httputil/reverseproxy.go;l=520;drc=960654be0c4ad7918376e2e1d47491c9bc7520e0 panics on the error setup in:
https://github.com/connectrpc/vanguard-go/blob/c6d89f4b3bbda0b6fa4737a9bb2c4425c4a23bfa/transcoder.go#L1530 which seems to be propagated in the `w.currentIsTrailer` branch when `err==nil` after it doesn't break but until it reaches the top of the loop again on the `w.err != nil` condition in `envelopingWriter`:  https://github.com/connectrpc/vanguard-go/blob/c6d89f4b3bbda0b6fa4737a9bb2c4425c4a23bfa/transcoder.go#L1328-L1362

Not sure if this is a right root cause or if it's something due to how `httputil.NewSingleHostReverseProxy` does buffering. As workaround I'm now catching the panic in the vanguard handler which seems to work ok for the client/server but is probably not good long term :)

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,5 +1,3 @@
-run:
-  skip-dirs-use-default: false
 linters-settings:
   errcheck:
     check-type-assertions: true
@@ -52,6 +50,7 @@ linters:
     - wrapcheck         # don't _always_ need to wrap errors
     - wsl               # generous whitespace violates house style
 issues:
+  exclude-dirs-use-default: false
   exclude:
     # Don't ban use of fmt.Errorf to create new errors, but the remaining
     # checks from err113 are useful.

--- a/regress_test.go
+++ b/regress_test.go
@@ -1,0 +1,135 @@
+// Copyright 2023-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vanguard_test
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+	"net/url"
+	"testing"
+
+	"buf.build/gen/go/connectrpc/eliza/connectrpc/go/connectrpc/eliza/v1/elizav1connect"
+	elizav1 "buf.build/gen/go/connectrpc/eliza/protocolbuffers/go/connectrpc/eliza/v1"
+	"connectrpc.com/connect"
+	"connectrpc.com/vanguard"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+// Reproducer test case provided in https://github.com/connectrpc/vanguard-go/issues/148
+func TestIssue148(t *testing.T) {
+	t.Parallel()
+
+	// GRPC-web server (process 1)
+	mux := http.NewServeMux()
+	mux.Handle(elizav1connect.NewElizaServiceHandler(&handler{t: t}))
+	grpcwebServer := httptest.NewUnstartedServer(
+		h2c.NewHandler(mux, &http2.Server{}),
+	)
+	grpcwebServer.EnableHTTP2 = true
+	grpcwebServer.Start()
+	t.Cleanup(grpcwebServer.Close)
+
+	// Vanguard based proxy (process 2)
+	webURL, err := url.Parse(grpcwebServer.URL)
+	require.NoError(t, err)
+	proxy := httputil.NewSingleHostReverseProxy(webURL)
+	proxy.FlushInterval = -1 // shouldn't be necessary
+	schema, err := protoregistry.GlobalFiles.FindDescriptorByName("connectrpc.eliza.v1.ElizaService")
+	require.NoError(t, err)
+	transcoder, err := vanguard.NewTranscoder(
+		[]*vanguard.Service{vanguard.NewServiceWithSchema(schema.(protoreflect.ServiceDescriptor), proxy)},
+		vanguard.WithDefaultServiceOptions(
+			vanguard.WithTargetProtocols(vanguard.ProtocolGRPCWeb),
+			vanguard.WithTargetCodecs(vanguard.CodecProto),
+		),
+		vanguard.WithUnknownHandler(proxy),
+	)
+	require.NoError(t, err)
+	vanguardServer := httptest.NewUnstartedServer(
+		h2c.NewHandler(
+			transcoder,
+			&http2.Server{},
+		),
+	)
+	vanguardServer.EnableHTTP2 = true
+	vanguardServer.Start()
+	t.Cleanup(vanguardServer.Close)
+
+	clientCases := map[string][]connect.ClientOption{
+		"proto": {connect.WithGRPC()},
+		"json":  {connect.WithGRPC(), connect.WithProtoJSON()},
+	}
+	for caseName := range clientCases {
+		caseName := caseName
+		clientOpts := clientCases[caseName]
+		t.Run(caseName, func(t *testing.T) {
+			t.Parallel()
+
+			// grpc client using h2c (process 3)
+			h2cClient := &http.Client{
+				Transport: &http2.Transport{
+					AllowHTTP: true,
+					DialTLSContext: func(ctx context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
+						return (&net.Dialer{}).DialContext(ctx, network, addr)
+					},
+				},
+			}
+
+			client := elizav1connect.NewElizaServiceClient(h2cClient, vanguardServer.URL, clientOpts...)
+			unaryResp, err := client.Say(context.Background(), connect.NewRequest(&elizav1.SayRequest{Sentence: "foo"}))
+			require.NoError(t, err)
+			require.Equal(t, "echo foo", unaryResp.Msg.Sentence)
+
+			serverStream, err := client.Introduce(context.Background(), connect.NewRequest(&elizav1.IntroduceRequest{Name: "foo"}))
+			require.NoError(t, err)
+			var responses []string
+			for serverStream.Receive() {
+				responses = append(responses, serverStream.Msg().Sentence)
+			}
+			require.NoError(t, serverStream.Err())
+			require.Equal(t, []string{"hi", "hi again"}, responses)
+		})
+	}
+}
+
+type handler struct {
+	elizav1connect.UnimplementedElizaServiceHandler
+	t testing.TB
+}
+
+func (h *handler) Say(_ context.Context, req *connect.Request[elizav1.SayRequest]) (*connect.Response[elizav1.SayResponse], error) {
+	if req.Peer().Protocol != connect.ProtocolGRPCWeb {
+		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("only accepts grpc-web, got %q", req.Peer().Protocol))
+	}
+	return connect.NewResponse(&elizav1.SayResponse{Sentence: "echo " + req.Msg.Sentence}), nil
+}
+
+func (h *handler) Introduce(_ context.Context, req *connect.Request[elizav1.IntroduceRequest], resp *connect.ServerStream[elizav1.IntroduceResponse]) error {
+	if req.Peer().Protocol != connect.ProtocolGRPCWeb {
+		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("only accepts grpc-web, got %q", req.Peer().Protocol))
+	}
+	_ = resp.Send(&elizav1.IntroduceResponse{Sentence: "hi"})
+	_ = resp.Send(&elizav1.IntroduceResponse{Sentence: "hi again"})
+	return nil
+}

--- a/transcoder.go
+++ b/transcoder.go
@@ -35,6 +35,10 @@ import (
 	"google.golang.org/protobuf/types/dynamicpb"
 )
 
+var (
+	errFinalDataAlreadyWritten = fmt.Errorf("final RPC response data already written: %w", context.Canceled)
+)
+
 // Transcoder is a Vanguard handler which acts like a router and a middleware. It transforms
 // all supported input protocols (Connect, gRPC, gRPC-Web, REST) into a protocol that the
 // service handlers support. It can do simple routing based on RPC method name, for simple
@@ -1222,8 +1226,7 @@ func (w *responseWriter) reportEnd(end *responseEnd) {
 	}
 	w.flusher.Flush()
 	// response is done
-	w.op.cancel()
-	w.err = context.Canceled
+	w.err = errFinalDataAlreadyWritten
 }
 
 func (w *responseWriter) flushHeaders() {
@@ -1311,7 +1314,7 @@ type envelopingWriter struct {
 	trailerIsCompressed bool
 }
 
-func (w *envelopingWriter) Write(data []byte) (int, error) {
+func (w *envelopingWriter) Write(data []byte) (n int, err error) {
 	w.maybeInit()
 	if w.err != nil {
 		return 0, w.err
@@ -1359,6 +1362,10 @@ func (w *envelopingWriter) Write(data []byte) (int, error) {
 			err := w.handleTrailer()
 			if err != nil {
 				return written, err
+			}
+			if len(data) == 0 {
+				// All done
+				return written, nil
 			}
 		} else {
 			// flush after each message and reset for next envelope
@@ -1527,7 +1534,7 @@ func (w *envelopingWriter) handleTrailer() error {
 	}
 	end.wasCompressed = w.trailerIsCompressed
 	w.rw.reportEnd(&end)
-	w.err = errors.New("final data already written")
+	w.err = errFinalDataAlreadyWritten
 	return nil
 }
 
@@ -1547,6 +1554,7 @@ type transformingWriter struct {
 	expectingBytes  int
 	writingEnvelope bool
 	latestEnvelope  envelope
+	trailerWritten  bool
 }
 
 func (w *transformingWriter) Write(data []byte) (n int, err error) {
@@ -1583,7 +1591,7 @@ func (w *transformingWriter) Write(data []byte) (n int, err error) {
 		w.buffer.Write(data[:remainingBytes])
 		written += remainingBytes
 		data = data[remainingBytes:]
-		if w.writingEnvelope {
+		if w.writingEnvelope { //nolint:nestif
 			var envBytes envelopeBytes
 			_, _ = w.buffer.Read(envBytes[:])
 			var err error
@@ -1606,6 +1614,10 @@ func (w *transformingWriter) Write(data []byte) (n int, err error) {
 			if err := w.flushMessage(); err != nil {
 				w.rw.reportError(err)
 				return written, err
+			}
+			if w.trailerWritten && len(data) == 0 {
+				// All done.
+				return written, nil
 			}
 			w.expectingBytes = envelopeLen
 			w.writingEnvelope = true
@@ -1651,7 +1663,8 @@ func (w *transformingWriter) flushMessage() error {
 		}
 		end.wasCompressed = w.latestEnvelope.compressed
 		w.rw.reportEnd(&end)
-		w.err = errors.New("final data already written")
+		w.err = errFinalDataAlreadyWritten
+		w.trailerWritten = true
 		return nil
 	}
 
@@ -1748,7 +1761,7 @@ type noResponseBodyWriter struct {
 }
 
 func (c noResponseBodyWriter) Write([]byte) (int, error) {
-	return 0, errors.New("final data already written")
+	return 0, errFinalDataAlreadyWritten
 }
 
 func (c noResponseBodyWriter) Close() error {

--- a/transcoder.go
+++ b/transcoder.go
@@ -1554,7 +1554,6 @@ type transformingWriter struct {
 	expectingBytes  int
 	writingEnvelope bool
 	latestEnvelope  envelope
-	trailerWritten  bool
 }
 
 func (w *transformingWriter) Write(data []byte) (n int, err error) {
@@ -1615,7 +1614,7 @@ func (w *transformingWriter) Write(data []byte) (n int, err error) {
 				w.rw.reportError(err)
 				return written, err
 			}
-			if w.trailerWritten && len(data) == 0 {
+			if w.latestEnvelope.trailer && len(data) == 0 {
 				// All done.
 				return written, nil
 			}
@@ -1664,7 +1663,6 @@ func (w *transformingWriter) flushMessage() error {
 		end.wasCompressed = w.latestEnvelope.compressed
 		w.rw.reportEnd(&end)
 		w.err = errFinalDataAlreadyWritten
-		w.trailerWritten = true
 		return nil
 	}
 

